### ISSUE TITLE
OpenMinTeD SSH UC Hackathon

### ISSUE THREAD

**Issue by reckart** [openminted/Open-Call-Discussions#6]
I have deployed a component and tried to run it on the platform. The result of the operation is listed as "FAILED", but I have no idea why. How can one get access to the log output?

![2018-04-08_14-37-41](https://user-images.githubusercontent.com/1410238/38470535-9d72b210-3b3a-11e8-9b9f-18c1d23fe0d5.png)

Instance: test.openminted.eu

**Comment by galanisd** [comment]
> I have deployed a component and tried to run it on the platform. 

For an application you can directly run it after you registration. If it is a component this is not possible.

**Comment by reckart** [comment]
> For an application you can directly run it after you registration. If it is a component this is not possible.

I know. I have built a workflow which makes use of the component that I had deployed (cf. : #7)

* omtdImporter
* PdfReader
* OpenNlpSegmenter
* VariableMentionDisambiguator

**Comment by reckart** [comment]
FYI @azielinskiACC

**Comment by galanisd** [comment]
OK
I had a look into Galaxy. 
VariableMentionDisambiguator is a UIMA component 
with the following coordinates

eu.openminted.uc-tdm-socialsciences
ss-variable-detection
1.0.1-SNAPSHOT

It is available on 
Maven Central ? 
zoidberg public snapshots? 
OMTD repo? -> the executor that we have does not look there.

Also the workflow is created in OMTD Workflow Editor instance of Galaxy. Then OMTD Registry
copies it OMTD Workflow Execution instance of Galaxy. Do you know the name of the workflow 
so I can check if it is there?

**Comment by reckart** [comment]
> OMTD repo? -> the executor that we have does not look there.

It is in the OMTD SNAPSHOTs repo. The registry seems to be able to resolve artifacts from there. Would it be possible to ensure that the executors and the registry use the same sets of repos to look up components, best also in the same order.

The workflow URL is: https://test.openminted.eu/landingPage/application/c58d1986-690e-40b9-b408-f649443c7d33

**Comment by galanisd** [comment]
> It is in the OMTD SNAPSHOTs repo. The registry seems to be able to resolve artifacts from there. Would it be possible to ensure that the executors and the registry use the same sets of repos to look up components, best also in the same order.

Until now it was not required. Added it on my TO-DO list.

> The workflow URL is: https://test.openminted.eu/landingPage/application/c58d1986-690e-40b9-b408-f649443c7d33

Downloaded the metadata record from Registry (attached). The workflow name is 
0931730980607790@openminted.eu 13865a76-613b-475a-88bf-4af5357b9263

I downloaded it from Galaxy executor (also attached). It is empty, no steps. Probably this is why
it fails. It seems a Registry issue.

[rec.zip](https://github.com/openminted/Open-Call-Discussions/files/1888007/rec.zip)

**Comment by reckart** [comment]
I'll try building a new one.

**Comment by galanisd** [comment]
Ok. Please sent me the landing page as you did with previous one. I will download the metadata record
find the Galaxy workflow and check if it is OK. If it is not we have to inform Antonis.

**Comment by reckart** [comment]
Ok, I have created a new one. This time, it is not empty when I re-open it in the workflow editor:

https://test.openminted.eu/landingPage/application/89d5e9ea-32fb-45f7-bf00-1fe466e33c4f

![2018-04-08_20-02-53](https://user-images.githubusercontent.com/1410238/38473561-4b386160-3b68-11e8-97aa-e672b194e953.png)

However, it still fails:

![2018-04-08_20-06-45](https://user-images.githubusercontent.com/1410238/38473569-672cc91a-3b68-11e8-88ca-4fd3549f2893.png)

@azielinskiACC @galanisd note that I have pasted a full multi-line XML file into the parameter **variableSpecification** - not sure if that could cause a problem. Aside from the XML getting a bit sqashed down when pasting it into the input field, it seemed ok in the Galaxy editor.

```
<?xml version="1.0" encoding="UTF-8"?>
<variables>
   <variable v_id="140" correct="YesNo">
       <v_label>INGLEHART-INDEX </v_label>
       <v_topic>Political attitudes and participation</v_topic>
       <v_question> What are your political priorities? </v_question>
       <v_subquestion> </v_subquestion>
       <v_answer a_id="1">Postmaterialist</v_answer>
       <v_answer a_id="2">Postmaterialist mixed-type</v_answer>
       <v_answer a_id="3">Materialist mixed-type</v_answer>
       <v_answer a_id="4">Materialist</v_answer>
       <v_answer a_id="5">Don't know</v_answer>
       <v_answer a_id="99">No answer</v_answer>
   </variable>
</variables>
```

The other thing is that the component should try to download a model from the OMTD Maven repo. That means it must have network access to that repo.

```
	<groupId>eu.openminted.uc-tdm-socialsciences</groupId>
	<artifactId>ss-variable-detection-model-disambiguation-en-ss</artifactId>
	<version>20180406.1</version>
```

Hm... that said, it might actually try to download the model from the wrong repo (i.e. the DKPro Core repo instead of the OMTD repo...). That is something I need to look into locally.

**Comment by reckart** [comment]
Opened an issue regarding model-auto-downloads here: https://github.com/openminted/omtd-component-executor/issues/1

**Comment by galanisd** [comment]
Yes now it not empty. 
The workflow is this 0931730980607790@openminted.eu 3c6c03b5-9a04-41bb-996a-a2cd536c7ace

I see a the following error in the logs workflow-service which is the module that call Galaxy. 

 --- [     Thread-625] e.o.w.service.WorkflowServiceImpl        : Unable to locate workflow: 0931730980607790%40openminted.eu+3c6c03b5-9a04-41bb-996a-a2cd536c7ace

Maybe it has to do with the name of the workflow. It contains spaces and a "@"  which are escaped at some point.
@courado @greenwoodma @antleb

**Comment by reckart** [comment]
@galanisd does OMTD set the document language in the PdfReader or do you expect that the language is set on all components in the workflow individually?

**Comment by reckart** [comment]
Ok. I have:
- fixed a bug in the OMTD-SHARE Maven plugin causing descriptors to be missing from the JARs
- changed the scope of the model from "test" to "compile" to avoid the component having to auto-download it
- uploaded the variable disambiguator component again; as far as I can tell the registry really picked up the latest SNAPSHOT version (https://test.openminted.eu/landingPage/component/104d0ca9-8d4f-4719-a337-9ecaa0a331e3)
- built a new workflow (https://test.openminted.eu/landingPage/application/60b9b72d-3e6d-4e5e-906d-766978cdf7c4)
- this time ensuring that I set the language "en" in the PdfReader so that the variable disambiguator component knows which model to use (language "en", variant "ss")

Then I tried running the workflow again on the variable test corpus that @azielinskiACC has published on the platform.

Still, I get a failure again.

Any idea what could be the reason now?

**Comment by galanisd** [comment]
I assume that again the workflow-service fails to call the workflow that was created @ Galaxy executor. As I said above probably the reason is the name of the workflow.

**Comment by greenwoodma** [comment]
I've just pushed a fix for this that should URL decode the workflow name before looking for it in Galaxy. This should get built and pushed to beta automatically but won't end up on test until someone manually pulls in the latest workflow service code.

**Comment by courado** [comment]
I have also added the error message supplied from the workflow service under the My Operations page

**Comment by reckart** [comment]
@courado great! :)

![2018-04-11_11-45-27](https://user-images.githubusercontent.com/1410238/38624190-e3d9652e-3d7d-11e8-82bd-eef099bc9a14.png)

I just tried running the workflow again, but it fails being unable to locate the named workflow.

Could somebody please push @greenwoodma `s fix to test.openminted.eu?

**Comment by greenwoodma** [comment]
@reckart is it not possible to rename the workflow to avoid the bug until the fix is pushed to test?

**Comment by reckart** [comment]
@greenwoodma how do I do that? The workflow editor only has a "save" button, not a "rename" or "save as" button as far as I remember.

**Comment by galanisd** [comment]
I think that the only way to do that is 

a. rename the workflow in Galaxy
b. download the metadata record of the app. delete it from the registry
c. upload an updated metadata record with the new workflow name.

**Comment by greenwoodma** [comment]
@reckart hmmm I thought the name of the workflow came from the name you gave the app in the registry UI, but maybe not, or maybe you can't change it there either. Certainly the workflow editor just gets passed the name from the platform it doesn't generate it.

**Comment by reckart** [comment]
Well, the name I have given to the workflow in the registry UI is "Simple Variable Disambiguation Example (English)". `0931730980607790@openminted.eu 3c6c03b5-9a04-41bb-996a-a2cd536c7ace` looks like an auto-generated ID over which I probably do not have control. My guess would be that it is a representation of the user-id concatenated with some other ID...

**Comment by greenwoodma** [comment]
What's weird is that if all workflow IDs are generated the same way then how have we ever run a workflow as we'd have hit this issue every time? I'm seriously confused by this one.

**Comment by reckart** [comment]
Apparently one **can** edit the workflow name in Galaxy by clicking on the pre-generated name, entering a new value and pressing ENTER. I did that (see screenshot).

![2018-04-11_12-08-31](https://user-images.githubusercontent.com/1410238/38625553-1e0c154a-3d81-11e8-97e7-f4ffa57b3066.png)

However, when I press "save" now, nothing happens. Odd...

Ok, when I go back to "My applications" and re-open the workflow in the editor, I can see that the name I put is still there, so I guess the "save" must have worked.

I wonder what happens if I created a second workflow by the same name...

Anyway, running the now re-named workflow still gives me the same message:

```
Failed 
Unable to locate named workflow
```

@courado the "My operations" view has a date, but not a time stamp. It would be great if we could also see the submission and possibly completion times of the execution there.

**Comment by galanisd** [comment]
@greenwoodma 

Workflow names @ Galaxy are not generated with the same way. 
* Some test Galaxy workflows were named by me (manually). I am calling them programmatically in our tests (via workflow-service).
* The Galaxy workflows that are created automatically from OMTD Registry and correspond to ready to use OMTD applications (e.g. Chebi app) seem to have valid names.
* The applications that are created in Galaxy editor and then ingested in OMTD Registry seem to have this problem.
@courado Please have a look on it.

Also workflow ID is a different thing that workflow name. For each workflow name there is an internal unique workflow ID; the one you retrieve in workflow-service from Galaxy so that you initiate a workflow execution.

**Comment by galanisd** [comment]
@greenwoodma 

> Apparently one can edit the workflow name in Galaxy by clicking on the pre-generated name, entering a new value and pressing ENTER. I did that (see screenshot).

I assume that this shouldn't be allowed and should be hidden as some other things @ Galaxy Editor.

**Comment by greenwoodma** [comment]
@galanisd yep, that most definitely shouldn't be allowed. I'll add it to the list of things I need to fix.

**Comment by galanisd** [comment]
OK great! It is not a blocking issue but this is 

> The applications that are created in Galaxy editor and then ingested in OMTD Registry seem to have this problem.

@courado

**Comment by reckart** [comment]
> OK great! It is not a blocking issue but this is
>
>> The applications that are created in Galaxy editor and then ingested in OMTD Registry seem to have this problem.

Indeed. We cannot proceed in the SSH UC (WP9) due to this issue at the moment.

@azielinskiACC

**Comment by courado** [comment]
Ok so:
- The sync of the galaxy editor and executor is managed through the registry.
- The workflow name is `user id + UUID` that is generated when you create a new application with the workflow editor.
- I cannot use the application name since we want it to be unique.
- If you change the name of the workflow, the `registry` passes the name generated upon creation to the `workflow service` and thus won't find it [_see save workflow save step 3_].
- Galaxy executor seems very slow when queried, `TODO` I need to add a loading screen when you press save.
--------------------------------------
When you save the workflow the steps are :
1. Load the workflow from the editor to the registry
2. Delete the workflow from the editor
3. If it was previously saved in the executor, delete it from the executor
4. Load the new workflow in the executor
----------------------------------
When you load a workflow :
If it already in the editor, show it else load it from the registry and show it

PS @galanisd:
1. The (un)escaping of the workflow name should be managed automatically from spring.
2. Clarify Workflow IDs are volatile and used internally, I do not send them to the workflow service, I just relay the workflow name which is the only thing remaining constant.

**Comment by greenwoodma** [comment]
Thanks @courado that all makes sense. Looks like the only issue is that the auto generated name doesn't get unencoded properly when passed into the workflow service hence it's looking for a workflow containing %40 instead of @ etc. My fix was to add a decode call inside the workflow service which should solve the problem once that code is deployed to test.

I agree the workflow IDs are volatile as they change everytime you export/import them into Galaxy, the only thing that's fixed is the workflow name so we do need that to be autog-enerated and unique so the current approach is great, I just need to fix the editor to stop people being able to change the name.

**Comment by galanisd** [comment]
> Thanks @courado that all makes sense. Looks like the only issue is that the auto generated name doesn't get unencoded properly when passed into the workflow service hence it's looking for a workflow containing %40 instead of @ etc. My fix was to add a decode call inside the workflow service which should solve the problem once that code is deployed to test.

@courado 
Is it possible to redeploy only workflow-service @ test so that we can check if Mark's fix works?
I think this is the easiest solution.

**Comment by greenwoodma** [comment]
I've just pushed a fix to the galaxy editor branch which stops you editing the workflow name from within the editor so that should appear next time test is fully updated with the latest versions of everything (assuming updating test includes the galaxy editor)

**Comment by reckart** [comment]
I am still getting `Failed - Unable to locate named workflow`.

Could anybody **please** update test.openminted.eu with the fixes that were discussed and implemented? 

I assume this will not be the last issue in the attempt of getting the SSH UC components running on the platform... and time is running out quickly.

**Comment by greenwoodma** [comment]
@reckart apparently test.openminted.eu has now been updated (sometime yesterday morning, plus again right now) so if you could try running your workflow again and see what happens?

**Comment by reckart** [comment]
Well... guess what: `Failed - Unable to locate named workflow`

**Comment by greenwoodma** [comment]
Damn, damn, damn and damn!

What I don't understand is that I tried to reproduce this myself by creating a new workflow through the galaxy editor and it worked. Having said that it looks as if it's worked because the workflow name doesn't contain an @ symbol like yours does.

Could you try creating a new workflow to see if that works (i.e. if something in the registry has changed the way it creates workflow names). The only other thing I can think of is that while test has been updated the workflow service is still the old one, but I'm not sure how to check that. @galanisd any ideas how we would check if the latest code had made it to test?

**Comment by greenwoodma** [comment]
also @reckart did you ever change the workflow name back after you managed to edit it? If not that would certainly screw things up

**Comment by reckart** [comment]
@greenwoodma I have no idea what the old name was.

**Comment by greenwoodma** [comment]
@reckart well that explains things then. Looking earlier in the issue I think it was

`0931730980607790@openminted.eu 13865a76-613b-475a-88bf-4af5357b9263`

if you can change it back to that then it might work, otherwise you need to create a new workflow to see if this has been fixed or not

**Comment by reckart** [comment]
![2018-04-13_10-53-06](https://user-images.githubusercontent.com/1410238/38738579-e2fb0680-3f08-11e8-8b05-c2baaabe856a.png)

I have rebuilt the workflow from scratch... now it is "running". Let's see if it terminates.

**Comment by greenwoodma** [comment]
@reckart is it worth closing this issue then, given it's now quite long and focused on the workflow name bug, and then opening another one if it fails with a different error?

**Comment by reckart** [comment]
@greenwoodma I have changed the issue title to "OpenMinTeD SSH UC Hackathon" - while the workflow name issue seems to be resolve now, the workflow still has not completed successfully.

**Comment by greenwoodma** [comment]
@reckart makes sense, just didn't know if you wanted a clean slate to report new issues, but renaming it for the hackathon makes sense

**Comment by reckart** [comment]
What is the usual time between status "running" and "completed"? 

@courado as a feature request to the registry:
* while the job is only scheduled and not running, maybe display "queued"
* show the rank of the job in the execution queue (maybe an estimated time to start) as a well as the time the job already spends in the queue

**Comment by reckart** [comment]
The workflow is trivial and the corpus is rather small, still the workflow is still in "running" state after three hours...

**Comment by greenwoodma** [comment]
@reckart how small is small? @antleb e-mailed me earlier about a slow running workflow. I'm beginning to wonder if the galaxy executor instance has been redeployed without the speed up fix we worked on for the issue that arose during the Paris meeting. I'm not entirely sure how to go about checking if that fix is in place or not -- will try and dig out the details for logging into the machine to check.

**Comment by reckart** [comment]
The corpus description says "one file"

https://test.openminted.eu/landingPage/corpus/9f4ebc21-aebe-4fb2-90c9-59bd189b9619

The corpus browser doesn't seem to work on that particular corpus.

**Comment by galanisd** [comment]
Someone has sent to executor ~2 hour ago  an 1 GB corpus. 

![1gb](https://user-images.githubusercontent.com/10739759/38749003-bba00284-3f58-11e8-9ef6-0cc10fc06ec9.png)

It will take ages.

**Comment by greenwoodma** [comment]
okay daft question then..... don't we allow parallel executions? I thought that was the point of the cloud backend?

**Comment by galanisd** [comment]
Did you try to download it?
https://test.openminted.eu/landingPage/corpus/9f4ebc21-aebe-4fb2-90c9-59bd189b9619

It is empty. So assume no data were feeded to the workflow. I think that in such cases workflow-service is not able to understand that processing has finished. Not sure.

@courado @antleb 
But why is empty?

**Comment by galanisd** [comment]
> okay daft question then..... don't we allow parallel executions? 

yes we allow parallel executions.

> I thought that was the point of the cloud backend?

yes..I don't disagree with you.
I just said that the specific one will take ages to finish and that Galaxy 
will be somehow overloaded when the results will be downloaded.

**Comment by greenwoodma** [comment]
@galanisd ah I see, and yes I would imagine Galaxy would be a bit of a bottleneck if it's trying to bring 1GB of docs into the history and then transfer it out to the workflow service

**Comment by reckart** [comment]
Where to put feature requests? I have some:

* ability to cancel operation
* ability to see if operation has actually started or is only queued
* see rank in queue
* sanity checking on corpus upload, i.e. reject empty corpora
* automatically detect if uploaded ZIP does **not** correspond to the required folder structure
* allow uploading of ZIPs containing only fulltext, no metadata or license files

**Comment by reckart** [comment]
![2018-04-13_21-16-06](https://user-images.githubusercontent.com/1410238/38762213-38f83056-3f60-11e8-9e99-13fdb009cc43.png)

Since the job from this morning never terminated (maybe because the corpus was empty), I have now uploaded a new (private) corpus, verified that it is not empty, and started a new job... let's see now...

**Comment by azielinskiACC** [comment]
several corpoa listed in OpenMinteD are actually empty (e.g. a brand new
testing corpus by penny; PLoSONE Corpus; Named Entity Recognition Corpus
for Social Science Publications, etc.)

However, download of  41 publications from OpenAIRE "OpenMinTeD subset of
OpenAIRE publications - Gkirtzou - Test2" was successful and I followed
their filestructure

and finally it worked: So try Variable Detection Corpus Test - 3 | Variable
Detection Corpus

https://test.openminted.eu/landingPage/corpus/e0d6b9b1-9b24-46aa-81db-b25fc37e85b0


Am Sa., 14. Apr. 2018 um 02:19 Uhr schrieb Richard Eckart de Castilho <
notifications@github.com>:

**Comment by galanisd** [comment]
> uploaded a new (private) corpus, verified that it is not empty, and started a new job... let's see now...

I tried to guess which workflow is executed in Galaxy for the job that you described..... from the screenshot that you sent and by searching Registry etc. It seems that it is a one PDF corpus (DAMECL.pdf?). Correct? No sure..just guessed. If I guessed correctly it seems that it got stuck in the first step (PDFReader). This is weird because I tried it myself with this component and worked.  

I think that we will speed up the process (of debugging) if you make the corpus and the application public and send me the landing pages. Then I will be able to test this job/execution and check everything in the background (Galaxy, Mesos nodes, worfklow-service). Otherwise it will take ages to find what is happening.

Dimitris.

**Comment by galanisd** [comment]
@reckart 

Also, I have just opened one workflow @ Galaxy executor. It has 4 steps and the last one is 
VariableMentionDisambiguator; probably it is one of those you created for testing. I noticed that the patterns parameter in PDFReader  step is not set. I think that I usually set it to something like "[+]**/*.pdf"  otherwise component execution fails.

Dimitris

**Comment by reckart** [comment]
> I think that I usually set it to something like "[+]**/*.pdf" otherwise component execution fails.

How did you manage to open a workflow and inspect the settings? When I open my workflow, I cannot see/edit the settings due to https://github.com/openminted/Open-Call-Discussions/issues/10

**Comment by galanisd** [comment]
I open Galaxy editor directly not through OMTD Registry.
There I can select a workflow from the respective list and download (export) it and or edit it.
I will send you some more info in a few minutes.

**Comment by reckart** [comment]
@galanisd I have rebuilt the workflow, this time with the patterns set on the PdfReader. Started it like 10 minutes ago. Didn't terminate so far.

![2018-04-14_10-58-45](https://user-images.githubusercontent.com/1410238/38769043-f5ad80da-3fd2-11e8-9836-be183dea7655.png)

**Comment by galanisd** [comment]
* yes but the 3 first steps were successfully executed; here the output of step 3.
[component_output.zip](https://github.com/openminted/Open-Call-Discussions/files/1910454/component_output.zip)

* Step 4 failed because of the variableSpecification parameter. 
See below the log messages from Step 4 container...The code that Galaxy automatically generates
from a wrapper has a syntax error. Probably because variableSpecification value is an XML.  

/srv/galaxy/database/jobs_directory/000/565/tool_script.sh: line 9: syntax error near unexpected token `<'
/srv/galaxy/database/jobs_directory/000/565/tool_script.sh: line 9: `mkdir tmp; cp /srv/galaxy/database/files/007/dataset_7405.dat tmp/DAMECL.pdf.xmi; cp /srv/galaxy/database/files/007/dataset_7406.dat tmp/typesystem.xml; Linux_runUIMA.sh eu.openminted.uc-tdm-socialsciences_ss-variable-detection_1.0.1-SNAPSHOT eu.openminted.uc.socialsciences.variabledetection.uima.VariableMentionDisambiguator -input tmp -output /srv/galaxy/database/jobs_directory/000/565/working/out/ -PmodelVariant='ss' -PvariableSpecification='<?xml version="1.0" encoding="UTF-8"?> <variables>    <variable v_id="140" correct="YesNo">        <v_label>INGLEHART-INDEX </v_label>        <v_topic>Political attitudes and participation</v_topic>        <v_question> What are your political priorities? </v_question>        <v_subquestion> </v_subquestion>        <v_answer a_id="1">Postmaterialist</v_answer>        <v_answer a_id="2">Postmaterialist mixed-type</v_answer>        <v_answer a_id="3">Materialist mixed-type</v_answer>        <v_answer a_id="4">Materialist</v_answer>        <v_answer a_id="5">Don't know</v_answer>        <v_answer a_id="99">No answer</v_answer>    </variable> </variables>' -PdisambiguateAllMentions='true' -PscoreThreshold='2.5' -PmaxMentions='3''

**Comment by galanisd** [comment]
Probably we shouldn't allow parameters with such values...We are making the whole thing to complicated. Galaxy has to support them..the executor has to support them...

An idea for this component (and for others) is to put some ready-to-use configuration files for variableSpecification in the jar...and to have a parameter that specifies which one should be used.

e.g variableSpecification=conf1

**Comment by reckart** [comment]
@galanisd The point is that we want users to use their own "variable specification", not a pre-defined one that we provide. Remember when I posted on the old mailing list about wanting to have the possibility for multi-line parameter values: this is the use-case.

**Comment by galanisd** [comment]
@reckart 

As it was mentioned here..
https://groups.google.com/forum/#!searchin/openminted-user-forum/xml$20text$20area%7Csort:date/openminted-user-forum/uYKBZCmqBX0/6ZijK7XrAwAJ 

escaping probably will solve the issue... still believe that we make things complicated. 
Needs some thought..

Dimitris

**Comment by galanisd** [comment]
> Remember when I posted on the old mailing list about wanting to have the possibility for multi-line parameter values: this is the use-case.

See above..I remembered it.

**Comment by reckart** [comment]
In the present situation, I believe anything else **but** allowing the users of the component to provide their own XML specification would make things more complicated.

Possible options:

* **user uploads a "model" which contains the specification and then uses that model with the component** -- IMHO doesn't work because: 1) we are not there yet with the model support, 2) it makes it way more complicated to try out a configuration because now you not only have to create a workflow (and enter metadata) but also have to upload a new model (and enter metadata) every time you even just want to fix a typo in the variable specification
* **configuration embedded in the JAR** -- this fails the point because the user cannot customize the variable specification to their own task. In the given use-case, the variable specification is tied very strongly to the user's problem at hand. It is not a generic thing like a POS tagger. Being unable to provide their own variable specification would render the component pretty much useless for anybody than the original component provider. Customizing would mean that a user has to build a new JAR with a new variable specification, release it, register it, and then use it in a workflow. IMHO this is also not feasible
* **provide the specification as a parameter** -- This is what we currently do. We'd provide the user with some documentation as to how the variable specification XML needs to be written. The user can then create a workflow with this configuration and use it.

**Comment by galanisd** [comment]
One question

* What is the difference between the parameters that are specified in the variable specification XML with the component parameters. Is it a good idea to make them component parameters?

Some thoughts

* I assume that if someone wants to provide a default configuration then a escaped XML should be put in the respective parameter element of OMTD-SHARE xml. We already do something similar e.g default value &lt;built-in&gt; escaped..

* If you provide an escaped XML in Galaxy editor (as you would do in OMTD-SHARE) for variable specification probably it will work. OR not? 

* Some ready-to-use configurations which are used very often could be included in the jar...
it will make life easier. But probably you are right that "providing the specification as a parameter" is the best solution ... but still thinking it.   

thanks!

**Comment by reckart** [comment]
> What is the difference between the parameters that are specified in the variable specification XML with the component parameters. Is it a good idea to make them component parameters?

Here is an example of the XML:

```
<?xml version="1.0" encoding="UTF-8"?>
<variables>
   <variable v_id="140" correct="YesNo">
       <v_label>INGLEHART-INDEX </v_label>
       <v_topic>Political attitudes and participation</v_topic>
       <v_question> What are your political priorities? </v_question>
       <v_subquestion> </v_subquestion>
       <v_answer a_id="1">Postmaterialist</v_answer>
       <v_answer a_id="2">Postmaterialist mixed-type</v_answer>
       <v_answer a_id="3">Materialist mixed-type</v_answer>
       <v_answer a_id="4">Materialist</v_answer>
       <v_answer a_id="5">Don't know</v_answer>
       <v_answer a_id="99">No answer</v_answer>
   </variable>
</variables>
```

If you boil it down to the essential that is actually used by the component, it is basically a list of key/value pairs:

* key: 140
* value: concatenation of `v_label`, `v_topic`, `v_question`, `v_subquestion`, `v_answer`elements

It would be possible to have a parameter which accepts key-value pairs encoded in some non-XML format, but IMHO it wouldn't make much of a different technically. It would still be a long multi-line parameter and there would still need to be some encoding because the users could put in arbitrary text for the value.

> I assume that if someone wants to provide a default configuration then a escaped XML should be put in the respective parameter element of OMTD-SHARE xml. We already do something similar e.g default value <built-in> escaped..

I didn't understand this one.

> If you provide an escaped XML in Galaxy editor (as you would do in OMTD-SHARE) for variable specification probably it will work. OR not?

Not sure what you mean? Do you ask the user to pre-escape the XML? IMHO that would make it again inconvenient to use.

**Comment by galanisd** [comment]
> Do you ask the user to pre-escape the XML? 
yeap. Not the best solution though.

Till now we didn't have the requirement to pass whole XML documents in a parameter. I am investigating  if and how this is supported in Galaxy ...

**Comment by reckart** [comment]
I think it is related to the configuration of the sanitizer in Galaxy: https://docs.galaxyproject.org/en/latest/dev/schema.html#tool-inputs-param-sanitizer

**Comment by galanisd** [comment]
yes I know. I have disabled sanitization otherwise Galaxy was filtering/removing some characters.

**Comment by gkirtzou** [comment]
@reckart You said, I quote 

> 
> What is the usual time between status "running" and "completed"?
> @courado as a feature request to the registry:
> - while the job is only scheduled and not running, maybe display "queued"
> -  show the rank of the job in the execution queue (maybe an estimated time to start) as a well as the time the job already spends in the queue
> 

As far as the 1st request is concerned, when a job starts, it is set as "Pending" from the registry, and it changes to "Running", when the workflow service starts running the job. You probably don't see this, as for the moment, there is no delay in the execution.

As far as the 2nd request, I don't know if there is a queue that could give you the rank position of a job. @greenwoodma or @courado would know better this part.

Also note that in the registry in each job, we keep the complete timestamp along with each status. I don't know why the interface just show the date. @courado ?

**Comment by greenwoodma** [comment]
I don't think there is a queue in Galaxy as such from which I could get a rank. As far as I can tell these things start almost straight away in Galaxy. I guess if there is a queue it will be in Chrons/Mesos to do with allocating cloud resources (i.e. waiting until there is some if we are at capacity)

**Comment by galanisd** [comment]
@reckart 

Created an updated docker image that contains an OMTD executor which can handle XML for parameter values. The only character that is not allowed and has to be escaped is single quote; it is used in Galaxy wrappers. 

e.g. variableSpecification value should be set to
```
<?xml version="1.0" encoding="UTF-8"?>
<variables>
    <variable v_id="140" correct="YesNo">
        <v_label>INGLEHART-INDEX </v_label>
        <v_topic>Political attitudes and participation</v_topic>
        <v_question> What are your political priorities? </v_question>
        <v_subquestion></v_subquestion>
        <v_answer a_id="1">Postmaterialist</v_answer>
        <v_answer a_id="2">Postmaterialist mixed-type</v_answer>
        <v_answer a_id="3">Materialist mixed-type</v_answer>
        <v_answer a_id="4">Materialist</v_answer>
        <v_answer a_id="5">Don'\''t know</v_answer>
        <v_answer a_id="99">No answer</v_answer>
    </variable>
</variables>
```

The updated image has been pushed ....
The only thing that you have to change is 

```
<v_answer a_id="5">Don't know</v_answer>
```  
to 

```
<v_answer a_id="5">Don'\''t know</v_answer>
```

and then you can retest.

If it works I will then try with Galaxy sanitizer 
https://docs.galaxyproject.org/en/latest/dev/schema.html#tool-inputs-param-sanitizer
to escape single quote automatically. 
To (fully) test this we have to 
* update galaxy wrapper generation 
* redeployment of the Registry
* delete and re-register your component. 

OR
* manually change wrapper for your tool with correct sanitization
* delete and re-register your component. 

One step at the time.

**Comment by galanisd** [comment]
@reckart 

Any updates? Did you try it?

**Comment by reckart** [comment]
@galanisd could only test today...

Just tried. I reopened the workflow linked above and since I cannot edit the configuration of existing components, I removed the variable disambiguation component, re-added it and then configured it. Basically, I simply removed the last two `v_answer` from the example XML.

Then I tried again running. Since there is only one document and a mini pipeline, it should complete quickly. But it still doesn't seem to terminate:

![2018-04-19_02-09-36](https://user-images.githubusercontent.com/1410238/38964441-2d2c2064-4377-11e8-9f15-6099a9c8f56c.png)

BTW. the "old" operations also still register as "running".

@courado feature request: 
* being able to cancel/delete operations
* at some point an operation should probably kill itself, some of mind are 4+ days old.

**Comment by reckart** [comment]
Btw. I have also registered the Keyword Assignment component now and try to run it on a single document corpus. This comment is mainly for documenting when I started it since this info is not shown in the operations screen. The pipeline is even more minimal than the disambiguation pipeline (no segmenter needed).

![2018-04-19_02-20-20](https://user-images.githubusercontent.com/1410238/38964692-76b71026-4378-11e8-999a-af335e576dd7.png)

@azielinskiACC

**Comment by greenwoodma** [comment]
So I've had a look at this issue of workflows running for ever and I think I've found the problem. I've just pushed a couple of fixes to the workflow service which should appear on beta shortly (not quite sure when they'll get pushed to test).

If you want the details read on.......

Essentially when a workflow runs we watch to see when the final step reaches the `ok` state (both the step and the underlying job). Unfortunately if an error occurs when running the workflow while this is captured and stored within the workflow service there wasn't an exception associated with the error (no exception was thrown as the error comes from checking the state not an exception). So while the internal object used for tracking progress within the workflow service recorded the failed state there was a problem when it came to communicating this to the registry. The JMS message doesn't contain a flag signifying the state of the workflow what it contains is an `error` field which should be filled with a message when an error occurs. The code in the workflow service filled this in using the message from the exception which had put the workflow into the failed state. Unfortunately in the case of a workflow failing because galaxy reported a state being in error there was no exception and so no message was returned. As such, while the workflow service knew that the workflow had failed the registry assumed it was still running and just sat there waiting for the next message from the workflow service which would never arrive. The fix involves never putting the internal object into the failed state without an associated exception, which means there is now always an error message (hopefully a useful one) which will be passed back to the registry.

@reckart I'm guessing your workflows are stuck in this situation. If you could send me the unique ID of the workflow (this is the long alphanumeric sequence next to the words "Workflow Canvas" at the top of the editor screen) then I can double check just to be certain. It won't help with working out why they failed, for that I'd need to look at the logs for the workflow service I think. @galanisd can you remind me the IP of the machine running the test instance of the workflow service?

**Comment by galanisd** [comment]
@reckart 

I created the same workflow with you; the Variable Dis. component is available in the Workflow editor.
I retested. Steps 1,2,3 were Ok...output as expected (checked Galaxy).

The Variable Dis. component fails while trying to download

de.tudarmstadt.ukp.dkpro.core#de.tudarmstadt.ukp.dkpro.core.variable-detection-model-disambiguation-en-default

part of the log attached.
[log.zip](https://github.com/openminted/Open-Call-Discussions/files/1927891/log.zip)

Locally in my laptop I do not have the same issue. I am trying to understand why...

**Comment by greenwoodma** [comment]
Would appear that the artifact isn't in any of the repos we look in.

**Comment by galanisd** [comment]
The 3 last steps of the workflow are DKPro UIMA components.
- PDFReader -> no model required

- OpenNLPSegmenter -> requires a model that was downloaded without issues.
e.g.
2018-04-19 11:52:32.555  INFO 50 --- [           main] d.t.u.d.c.a.r.ResourceObjectProviderBase : downloading http://zoidberg.ukp.informatik.tu-darmstadt.de/artifactory/public-model-releases-local/de/tudarmstadt/ukp/dkpro/core/de.tudarmstadt.ukp.dkpro.core.opennlp-model-token-en-maxent/20120616.1/de.tudarmstadt.ukp.dkpro.core.opennlp-model-token-en-maxent-20120616.1.jar ...
2018-04-19 11:52:32.896  INFO 50 --- [           main] d.t.u.d.c.a.r.ResourceObjectProviderBase :       [SUCCESSFUL ] de.tudarmstadt.ukp.dkpro.core#de.tudarmstadt.ukp.dkpro.core.opennlp-model-token-en-maxent;20120616.1!de.tudarmstadt.ukp.dkpro.core.opennlp-model-token-en-maxent.jar (340ms)

- Variable Dis. -> requires a model but fails to downloaded it.

Hmmm...

**Comment by reckart** [comment]
The model that the VarDis is using should be in the same repo as VarDis itself - however, in according to the logs, it tries to download the "default" variant, not the "ss" variant. I'm trying to check the workflow config again.

**Comment by galanisd** [comment]
I am checking the configuration for the repos in my laptop. I deleted the model but the when I run the script it is downloaded...

**Comment by galanisd** [comment]
I was using modelLocation not modelVariant. 
Corrected. I am retesting right now.

**Comment by reckart** [comment]
The models are here: https://repo.openminted.eu/content/repositories/releases/eu/openminted/uc-tdm-socialsciences/

**Comment by reckart** [comment]
> @reckart I'm guessing your workflows are stuck in this situation. If you could send me the unique ID of the workflow (this is the long alphanumeric sequence next to the words "Workflow Canvas" at the top of the editor screen) then I can double check just to be certain.

There are at least two ones stuck with jobs:

* 0931730980607790-9b9b1d64-fe3f-4de7-88ca-bf1f788e60f5
* 0931730980607790-bcdd2736-498c-48e9-b61d-352a043e7175

Btw. I can still edit the workflow name in the workflow editor.

**Comment by galanisd** [comment]
Got results.... :-)
![variabledis](https://user-images.githubusercontent.com/10739759/38991701-3fe9766e-43e7-11e8-8a65-124ee02a0b15.png)

Attached...

[bc4e4776-cc9c-47d1-bf28-0d9b5ab78c46.zip](https://github.com/openminted/Open-Call-Discussions/files/1928115/bc4e4776-cc9c-47d1-bf28-0d9b5ab78c46.zip)

I hope that is not an illusion...

![result](https://user-images.githubusercontent.com/10739759/38991722-5098b420-43e7-11e8-9c34-5a85b8c26d54.png)


- Still I have to check what happens with the repo configuration even though it seems to work right now.

- I do not understand why your workflow never completed.

**Comment by reckart** [comment]
@galanisd great news!!!

For curiosity: does it open in the Annotation Viewer?

**Comment by galanisd** [comment]
Nope ..... I think because the results are written in an "output" and not in an "annotations" folder. 
This happens because currently the metadata of the component are not passed to the workflow-service.

I might be wrong.
@greenwoodma @antleb @courado ?

**Comment by greenwoodma** [comment]
Yes, there is a redmine issue https://redmine.openminted.eu/issues/767 which I've just bumped.

**Comment by azielinskiACC** [comment]
So, finally. That's great. 
Was it possible to use a configuration file?
For NER I use the following https://test.openminted.eu/landingPage/application/2d3fc2aa-6f9b-4a5b-bd75-763a39b8b18b
Correct?

**Comment by reckart** [comment]
@azielinskiACC I cannot access the link above. Probably it is a private workflow in your account?

**Comment by galanisd** [comment]
I can ... 

![ner](https://user-images.githubusercontent.com/10739759/38997429-13fe0cfe-43f6-11e8-87fb-07b5839dd953.png)
however this metadata record seems to be for an image that I have created 10 months ago...
(Identifiers OMTD: DemoWF3SSHNER)

Back then there were no docker specs and we have create 5 apps (one of them was NER) in order

1. to do some demos
2. to experiment with Mesos/Galaxy and see what is required.

The respective image us not OMTD compliant; i.e. it does not follow the docker spec and it will not be executed in the current environment.

See also ...
https://github.com/openminted/Open-Call-Discussions/issues/1


Who is working on this image/app?

**Comment by reckart** [comment]
I'd have to look into the NER thing.

**Comment by galanisd** [comment]
If required please open a new issue (NER Hackathon)

**Comment by azielinskiACC** [comment]
For testing, it would be great to have the proper landing ID for all SS-A applications, since search on the OpenMinted Platform does not give any results.
Unfortunately, there are some 'empty' corpora I created and cannot be deleted and which might cause confusion (A known issue?) So please also let me know which data input files (=landing ID) I should use.

**Comment by reckart** [comment]
@azielinskiACC @galanisd since the "test.openminted.eu" platform is only for testing and may be reset again... does it make sense at all to use fixed IDs for corpora? Maybe better to have people upload own data or build a corpus using the search functionality.

The names of the SSH components on the other hand are rather stable. I'll run a release and then could publish them to the main platform (non-test).

**Comment by pennyl67** [comment]
@reckart Please note that @antleb is currently updating the main platform (services), so I wouldn't recommend adding anything there until we get notified. The idea is to use the services for all the testing etc., so it must be updated with all the fixes that the test platform has now.

**Comment by pennyl67** [comment]
Sorry, by "testing" I meant the evaluation of the tenders/hackathon

**Comment by greenwoodma** [comment]
@pennyl67 is @antleb updating services to the same as test is currently or to the latest version of the code? The plan was to update test daily since the WP7 call last week, but the workflow-service hasn't been updated in the last week so it's still not got all the bug fixes we've made this week (which is quite a few). The problem is that while I think those fixes all work as expected, I'd assumed they were being tested on test as that was being updated. Now I find it hasn't been, so it may be that we get an up to date services which is buggier than test.

**Comment by reckart** [comment]
@pennyl67 @azielinskiACC @galanisd @antleb WP9 also has to wrap up the "tutorial" material. My comment was related to what we can expect to later be able to find on the main platform and what not. Things we can find on the main platform can be in the tutorial, others maybe not. So IMHO it would make more sense to include the building of the corpus/uploading of documents into the tutorial, also the building of a workflow, but not the uploading of the components - we should be able to expect that the components we test now will exist on the main platform. Makes sense?

**Comment by pennyl67** [comment]
@greenwoodma I'm not sure - I didn't know this detail; trying to find out and i'll let you know
@reckart I've asked @antleb to not delete from the main platform anything before I check; the problem is that we need to clean up various test resources and empty corpora. And I will send out an email to all to check which resources they want us to keep. 
And yes, I understand what you're saying for the tutorials (everything happening at the same time!) - but how can you build a workflow without the components in the same platform? I can understand for the built/updated corpora.

**Comment by reckart** [comment]
We publish the SSH UC components to the test platform now for the preparation of the tutorial material. The same components will later be published to the main platform - once the main platform is ready.

**Comment by reckart** [comment]
My understanding is that for the VarDis and Keyword components, we are good now.

@azielinskiACC - do you agree? If yes, I would suggest to close this issue.

**Comment by azielinskiACC** [comment]
Yes - both components are running and the issue can be closed.

Am Fr., 20. Apr. 2018 um 14:03 Uhr schrieb Richard Eckart de Castilho <
notifications@github.com>:
